src/connectors/voice-search/connectVoiceSearch.js

```diff
@@ -0,0 +1,87 @@
+import {
+  checkRendering,
+  createDocumentationMessageGenerator,
+} from '../../lib/utils';
+import voiceSearchHelper from '../../lib/voiceSearchHelper';
+
+const withUsage = createDocumentationMessageGenerator({
+  name: 'voice-search',
+  connector: true,
+});
+
+export default function connectVoiceSearch(renderFn, unmountFn) {
+  checkRendering(renderFn, withUsage());
+
+  return (widgetParams = {}) => {
+    return {
+      init({ helper }) {
+        this._refine = (() => {
+          let previousQuery;
+          const setQueryAndSearch = query => {
+            if (query !== helper.state.query) {
+              previousQuery = helper.state.query;
+              helper.setQuery(query);
+            }
+            if (previousQuery !== undefined && previousQuery !== query) {
+              helper.search();
+            }
+          };
+
+          return setQueryAndSearch;
+        })();
+
+        this._rerender = () => {};
+
+        this._voiceSearchHelper = voiceSearchHelper({
+          onQueryChange: query => this._refine(query),
+          onStateChange: state => {
+            this._voiceListeningState = state;
+            this._rerender();
+          },
+        });
+
+        this._toggleListening = searchAsYouSpeak => {
+          const { isListening, start, stop } = this._voiceSearchHelper;
+          if (isListening()) {
+            stop();
+          } else {
+            start(searchAsYouSpeak);
+          }
+        };
+
+        renderFn(
+          {
+            query: helper.state.query,
+            isSupportedBrowser: this._voiceSearchHelper.isSupportedBrowser,
+            isListening: this._voiceSearchHelper.isListening,
+            toggleListening: this._toggleListening,
+            voiceListeningState: this._voiceListeningState,
+            widgetParams,
+          },
+          true
+        );
+      },
+      render({ helper }) {
+        this._rerender = () => {
+          this.render({ helper });
+        };
```

Comment:
This is rather tricky 😅 I understand why need it but I wonder if there is anyway to avoid it.
Did you find similar patterns in part of the code base that deal with the same kind of re-rendering issue?

Reply:
This is the same pattern as used for showMore

Reply:
We use en external function for the `render` in the [other](https://github.com/algolia/instantsearch.js/blob/develop/src/connectors/refinement-list/connectRefinementList.js#L359-L371) implementations. Here we heavily rely on `this` which doesn't help to follow how the code is actually executed.

Reply:
That looks clearer 👍